### PR TITLE
Update Product in Sales Order line should update the product in Sales Invoice Disposition

### PR DESCRIPTION
**Pull request by riadmen**
#17998

### CODE REVIEW

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/callout/C_OrderLine.java [hunk 1]
```diff
@@ -24,23 +24,43 @@
...
+import org.adempiere.exceptions.AdempiereException;
 import org.compiere.SpringContextHolder;
 
 @Callout(I_C_OrderLine.class)
 public class C_OrderLine
 {
 	private final IOrderLineBL orderLineBL = Services.get(IOrderLineBL.class);
+	private final IInvoiceCandDAO invoiceCandDAO = Services.get(IInvoiceCandDAO.class);
 	private final IDocumentLocationBL documentLocationBL = SpringContextHolder.instance.getBean(IDocumentLocationBL.class);
+	private static final String MSG_ERROR_INVOICE_CANDIDATE_IS_PROCESSED = "C_OrderLine.onProductChanged.Msg_Error_Invoice_Candidate_Is_Processed";
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/callout/C_OrderLine.java:44]: pls use AdMessageKey

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/callout/C_OrderLine.java [hunk 2]
```diff
@@ -24,23 +24,43 @@
...
 	@CalloutMethod(columnNames = { I_C_OrderLine.COLUMNNAME_M_Product_ID })
 	public void onProductChanged(final I_C_OrderLine orderLine)
 	{
+		final OrderLineId orderLineId = OrderLineId.ofRepoId(orderLine.getC_OrderLine_ID());
+
+		for (final I_C_Invoice_Candidate icRecord : invoiceCandDAO.retrieveInvoiceCandidatesForOrderLineId(orderLineId))
+		{
+			if (icRecord.isProcessed())
+			{
+				throw new AdempiereException("@" + MSG_ERROR_INVOICE_CANDIDATE_IS_PROCESSED+ "@");
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/callout/C_OrderLine.java:55]: pls use AdMessageKey and avoid "@" concatenation

Comment (teosarca) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/callout/C_OrderLine.java:55]: also IMHO i would introduce InvoiceCandBL.assertNotProcessedForOrderLineId

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/invoicecandidate/spi/impl/ProductFromOrderToInvoiceCandidateTest.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+		assertThat(invoiceCandidates).as("There is 1 invoice candidate").hasSize(1);
+
+		final I_C_Invoice_Candidate candidate = invoiceCandidates.get(0);
+		assertThat(candidate.getInvoiceRule()).isNotNull();
+		saveRecord(candidate);
+		// ReActivate
+		docActionBL.processEx(order, IDocument.ACTION_ReActivate, IDocument.STATUS_InProgress);
+		orderLine.setM_Product_ID(changedProductId.getRepoId());
+
+		olHandler.setOrderedData(candidate);
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/invoicecandidate/spi/impl/ProductFromOrderToInvoiceCandidateTest.java:265]: TBH if u call setOrderedData directly here.... i am not sure what are you exactly testing....

i.e. u could call that method directly without reactivating the order....

so what's exactly ur test scenario?

Comment (riadmen) [reply backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/invoicecandidate/spi/impl/ProductFromOrderToInvoiceCandidateTest.java:265]: - Like the WEB UI test:

1. Create an SO and complete it
2. IC is created with product 1
3. Reactivate the SO
4. Change the product in the OL
5. Compare the product in the OL and the IC

**Pull request merged**